### ISSUE TITLE
BCE Loss Compatibility Issue in the Colored MNIST Reproduction

### ISSUE THREAD

**Issue by Banabo-code** [cvlab-columbia/CT4Recognition#7]
Hello, thank you for releasing the code and pretrained models. We are trying to reproduce the Colored MNIST experiment and have encountered a compatibility issue related to the BCE loss.
In recent versions of PyTorch, F.binary_cross_entropy checks whether both the predictions and the targets are within the range [0,1]. However, your code applies : transforms.Normalize([0.5], [0.5]) to the input images, which transforms the image range from [0,1] to [-1,1]. At the same time, the VAE reconstruction is passed through a sigmoid, so recon_x is in [0,1]. Therefore, directly computing: F.binary_cross_entropy(recon_x, x) produces an error in recent PyTorch versions.
We suspect that the older version of BCE used in your original environment may have checked only the predictions, but not the target values. This could explain why the code ran successfully at that time. However, even if it runs, using targets in the range [-1,1] with BCE seems conceptually problematic.
To address this issue, we kept the Normalize transform and mapped the BCE target from [-1,1] back to [0,1] : target = (x + 1.0) / 2.0 
BCE = F.binary_cross_entropy(recon_x, target, reduction='sum')
We have attached our modified training script.

[fd_classifier_cmnist_train.py](https://github.com/user-attachments/files/31454549/fd_classifier_cmnist_train.py)

 The code now runs successfully, but the results are still substantially different from those reported in the paper. The causal method achieves approximately 38% accuracy, while the Vanilla method achieves only about 18%. In the paper, the reported accuracies are approximately 50% for the causal method and 39% for the Vanilla [method.]
Could you please help us with the following questions?
1. Which versions of PyTorch, torchvision, CUDA, and Python did you use?
2. In your original environment, did BCE actually allow target values in the range [-1,1]?
3. How should the training code be modified for recent PyTorch versions so that the preprocessing remains correct and the results can approach the reported 50% causal accuracy and 39% Vanilla accuracy?
4. Were there any other training configurations, random seeds, or data-processing details that were not included in the released code?
Thank you very much for your help.
